### PR TITLE
Alex

### PR DESCRIPTION
**Pull request by tmpr**

### CODE REVIEW

File: shared/constraints.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+from typing import Callable
+
+import numpy as np
+
+from simplex.simplex_problem import SimplexCallable
+
+
+@dataclass
+class Constraint:
+    c: Callable[[np.ndarray], np.ndarray]
```
Comment (rendstrasser) [inline shared/constraints.py:15]: typing: should return float

File: shared/constraints.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    is_equality: bool
+
+    def __call__(self, x: np.ndarray) -> np.ndarray:
```
Comment (rendstrasser) [inline shared/constraints.py:18]: typing: should return float

File: shared/constraints.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        """Check whether the constraint is active at point x, i.e. if c(x) == 0.
+
+        Args:
+            x: Point to evaluate the constraint at.
+        """
+        return self(x) == 0
+
+
+@dataclass
+class SimplexConstraint(Constraint):
```
Comment (rendstrasser) [inline shared/constraints.py:34]: not necessarilly simplex (also relevant for quadratic), maybe other name?
Maybe LinearConstraint?

File: shared/minimization_problem.py [hunk 4]
```diff
@@ -16,19 +16,27 @@ class MinimizationProblem:
 
     Args:
         f (Callable): The function (objective) we are trying to minimize.
-        constraints (list): List of constraints functions for the minimization problem.
-        solution (list): The solution(s) to the minimization problem.
-                             Might contain multiple if there are multiple local minimizers.
-        x0 (list): The starting point for the minimization procedure.
+        constraints (Sequence[Constraint]): Sequence of constraints.
+        solutions (Sequence[np.ndarray]): The solutions(s) to the minimization problem.
+                                         Might contain multiple if there are multiple local minimizers.
```
Comment (rendstrasser) [inline shared/minimization_problem.py:21]: My bad, not multiple solutions (atm) -> Sequence[float]

File: shared/minimization_problem.py [hunk 5]
```diff
@@ -39,41 +47,43 @@ def calc_hessian_at(self, x: np.ndarray) -> np.ndarray:
...
     # --- Constraint methods ---
 
-    def calc_constraints_at(self, x: np.ndarray) -> np.ndarray: 
-        return np.array([c.c(x) for c in self.constraints])
+    def calc_constraints_at(self, x: np.ndarray) -> np.ndarray:
+        return np.array([c(x) for c in self.constraints])
 
-    def calc_constraint_at(self, i: int, x: np.ndarray) -> float: 
-        return self.constraints[i].c(x)
+    def calc_constraint_at(self, i: int, x: np.ndarray) -> np.ndarray:
```
Comment (rendstrasser) [inline shared/minimization_problem.py:53]: float is returned

File: simplex/simplex_problem.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+from typing import Sequence
+
+import numpy as np
+
+from shared.constraints import SimplexConstraint
+from shared.minimization_problem import MinimizationProblem
+
+
+@dataclass
+class SimplexCallable:
```
Comment (rendstrasser) [inline simplex/simplex_problem.py:15]: naming: SimplexCallable sounds like the general function, not a constraint
also, same information relevant for quadratic programming

maybe LinearConstraintCallable or so?

File: simplex/simplex_problem.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    """Specific callable to SimplexProblem. Keeps A and b accessible."""
+    A: np.ndarray
```
Comment (rendstrasser) [inline simplex/simplex_problem.py:17]: actually, a single constraint only contains one a_i of A
same with b, it is only a single value b_i

We may need to expose methods on the problem to get the entire A and b

**Review by rendstrasser — CHANGES_REQUESTED** [review]

File: shared/constraints.py [hunk 8]
```diff
@@ -31,5 +29,28 @@ def is_active(self, x: np.ndarray) -> bool:
 
 
 @dataclass
-class SimplexConstraint(Constraint):
-    c: SimplexCallable
+class LinearCallable:
+    """Specific callable to SimplexProblem. Keeps A and b accessible."""
```
Comment (rendstrasser) [inline shared/constraints.py:33]: (minor) not necessarily SimplexProblem, will also be relevant for QuadraticProblem

File: simplex/simplex_problem.py [hunk 9]
```diff
@@ -2,27 +2,21 @@
...
 class SimplexProblem(MinimizationProblem):
-    constraints: Sequence[SimplexConstraint]
+    f: Callable[[np.ndarray], np.ndarray] = field(init=False)
+    constraints: Sequence[LinearConstraint]
+    c: np.ndarray
 
-    # Methods
+    def __post_init__(self):
+        self.f = lambda x: np.inner(self.c, x)
+        self.A, self.b = combine_linear([constraint.c for constraint in self.constraints])
```
Comment (rendstrasser) [inline simplex/simplex_problem.py:22]: maybe add A and b to SimplexProblem as well such that you clearly see that you can use them (with field(init=False))

**Review by rendstrasser — APPROVED** [review]

**Review by rendstrasser — APPROVED** [review]

**Comment by Tigxy** [comment]
Very good. Christian approved. (╯✧▽✧)╯

**Pull request merged**